rasa/core/policies/embedding_policy.py

```diff
@@ -295,116 +233,234 @@ def _load_params(self, **kwargs: Dict[Text, Any]) -> None:
+            multi_X,
+            multi_Y,
+            multi_labels,
+            test_size=self.evaluate_on_num_examples,
+            random_state=self.random_seed,
+            stratify=multi_labels,
+        )
+        X_train = np.concatenate([X_train, solo_X])
+        Y_train = np.concatenate([Y_train, solo_Y])
+        labels_train = np.concatenate([labels_train, solo_labels])
+
+        return (
+            SessionData(X=X_train, Y=Y_train, labels=labels_train),
+            SessionData(X=X_val, Y=Y_val, labels=labels_val),
+        )
 
-        return np.stack([self.encoded_all_actions] * dialogue_len)
+    @staticmethod
+    def _shuffle_session_data(session_data: "SessionData") -> "SessionData":
+        """Shuffle session data."""
 
+        ids = np.random.permutation(len(session_data.X))
+        return SessionData(
+            X=session_data.X[ids],
+            Y=session_data.Y[ids],
+            labels=session_data.labels[ids],
+        )
+
+    # tf helpers:
     # noinspection PyPep8Naming
-    def _create_tf_session_data(
-        self, domain: "Domain", data_X: np.ndarray, data_Y: Optional[np.ndarray] = None
-    ) -> SessionData:
-        """Combine all tf session related data into a named tuple"""
+    def _gen_batch(
+        self,
+        session_data: "SessionData",
+        batch_size: int,
+        batch_strategy: Text = "sequence",
+        shuffle: bool = False,
+    ) -> Generator[Tuple["np.ndarray", "np.ndarray"], None, None]:
+        """Generate batches."""
+
+        if shuffle:
+            session_data = self._shuffle_session_data(session_data)
+
+        if batch_strategy == "balanced":
+            num_examples = len(session_data.X)
+            unique_labels, counts_labels = np.unique(
+                session_data.labels, return_counts=True, axis=0
+            )
+            num_labels = len(unique_labels)
+
+            label_data = []
+            for label in unique_labels:
+                label_data.append(
+                    SessionData(
+                        X=session_data.X[session_data.labels == label],
+                        Y=session_data.Y[session_data.labels == label],
+                        labels=None,  # ignore new labels
+                    )
+                )
 
-        X, slots, previous_actions = self._create_X_slots_previous_actions(data_X)
+            data_idx = [0] * num_labels
+            num_data_cycles = [0] * num_labels
+            skipped = [False] * num_labels
+            new_X = []
+            new_Y = []
+            while min(num_data_cycles) == 0:
+                if shuffle:
+                    ids = np.random.permutation(num_labels)
+                else:
+                    ids = range(num_labels)
+
+                for i in ids:
+                    if num_data_cycles[i] > 0 and not skipped[i]:
+                        skipped[i] = True
+                        continue
+                    else:
```

Comment:
minor: else not needed

Reply:
I think we need it because `skipped[i]` might be `True` from previous cycle